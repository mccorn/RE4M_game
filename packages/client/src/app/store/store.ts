import { configureStore } from '@reduxjs/toolkit';
import { userReducer } from './slices/userSlice';
import { gameReducer } from './slices/gameSlice';

export const reducer = { user: userReducer, game: gameReducer };

declare global {
    interface Window {
        __PRELOADED_STATE__?: object;
    }
}

export const store = configureStore({
    reducer,
    // eslint-disable-next-line no-underscore-dangle
    preloadedState: window.__PRELOADED_STATE__,
});

// eslint-disable-next-line no-underscore-dangle
delete window.__PRELOADED_STATE__;

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
