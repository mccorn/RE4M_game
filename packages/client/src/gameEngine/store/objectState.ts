import { TPoint } from '../types/commonTypes';
import Trajectory from '../objects/trajectory';
import { DrawableObjectState } from '../core/drawableGameObject';

export enum GlobalGameState {
    Loaded,
    LevelLoading,
    LevelStarted,
    Paused,
    Resumed,
    LevelEnded,
    Ended,
}

// we draw ship in Flying, Shooting and Exploiding states
export enum LiveState {
    WaitForStart,
    Flying,
    Shooting,
    Exploiding,
    Dead,
}

export const GAME_EVENTS = {
    objectIsDead: 'objectIsDead',
};

export const EVENTS = {
    [GAME_EVENTS.objectIsDead]: new Event(GAME_EVENTS.objectIsDead),
};

export class ShipState extends DrawableObjectState {
    private liveState: LiveState;

    constructor(coordinates: TPoint, trajectory: Trajectory, liveState: LiveState) {
        super(coordinates, trajectory);
        this.liveState = liveState;
    }

    // todo check do we need +
    public isExploiding = () => this.liveState === LiveState.Exploiding;

    public isWaiting = () => this.liveState === LiveState.WaitForStart;

    public isFlying = () => this.liveState === LiveState.Flying;

    public isDead = () => this.liveState === LiveState.Dead;

    public setLiveState = (liveState: LiveState) => {
        if (liveState === LiveState.Dead && this.liveState !== LiveState.Dead) {
            window.dispatchEvent(EVENTS.objectIsDead);
        }

        this.liveState = liveState;
    };

    // 2 action change index and set dead
    public changeFrameIndex = (frameCount: number, shouldChangeFrame: boolean) => {
        if (this.isExploiding() && shouldChangeFrame) {
            this.frameIndex += 1;

            if (this.frameIndex >= frameCount) {
                this.setLiveState(LiveState.Dead);
            }
        }
    };

    public updateEnemyState = (time: number, frameCount: number, shouldChangeFrame: boolean) => {
        this.followTrajectory(time);

        if (this.isWaiting() && this.trajectory.shouldStartMoving(time)) {
            // if ship should starts moving set state to Flying
            this.setLiveState(LiveState.Flying);
        } else if (!this.isDead() && this.trajectory.movedOutOfGameField(time)) {
            // if ship flied out of canvas set state to Dead
            this.setLiveState(LiveState.Dead);
        }

        this.changeFrameIndex(frameCount, shouldChangeFrame);
    };
}

export class ShotState extends DrawableObjectState {
    private show: boolean;

    private startTime: number;

    constructor(coordinates: TPoint, trajectory: Trajectory, show: boolean, startTime: number) {
        super(coordinates, trajectory);
        this.show = show;
        this.startTime = startTime;
    }

    private hide = () => {
        this.show = false;
    };

    public isVisible = () => this.show;

    public changeFrameIndex = (shouldChangeFrame: boolean, frameCount: number) => {
        if (shouldChangeFrame) {
            this.frameIndex += 1;
            if (this.frameIndex >= frameCount) {
                this.frameIndex = 0;
            }
        }
    };

    public update = (time: number, shouldChangeFrame: boolean, frameCount: number) => {
        const { trajectory } = this;
        const deltaTime = time - this.startTime;
        if (trajectory && trajectory.shouldMove(deltaTime)) {
            this.setCoordinates(trajectory.getCoordinates(deltaTime));
        }
        this.changeFrameIndex(shouldChangeFrame, frameCount);

        if (trajectory.movedOutOfGameField(time)) {
            this.hide();
        }
    };
}
