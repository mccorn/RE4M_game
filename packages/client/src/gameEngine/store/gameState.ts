import PlayerShip from '../objects/ships/playerShip';
import params, { NEXT_SHIP_DELAY } from '../parameters/gameParameters';
import { TEnemyType } from '../types/commonTypes';
import GameLevels, { GameLevelList } from '../parameters/gameLevels';
import Trajectory from '../objects/trajectory';
import { GlobalGameState } from './objectState';
import EnemyShip from '../objects/ships/enemyShip';
import GameShot from '../objects/gameShot';

class GameState {
    public player: PlayerShip;

    public enemies: EnemyShip[] = [];

    public shots: GameShot[] = [];

    private currentLevel: GameLevelList;

    private state: GlobalGameState;

    public isGameRunning = false;

    constructor() {
        this.currentLevel = GameLevelList.Level1;
        this.player = this.initPlayer();
        this.state = GlobalGameState.Loaded;
    }

    // eslint-disable-next-line class-methods-use-this
    private initPlayer = () => new PlayerShip(params.PLAYER_COORDINATES);

    private initEnemies = () => {
        const ships: EnemyShip[] = [];
        const levelParams = GameLevels[this.currentLevel];
        Object.keys(levelParams.enemies).forEach(key => {
            const type = key as unknown as TEnemyType;
            const enemyLevelParams = levelParams.enemies[type];
            if (enemyLevelParams) {
                for (let i = 0; i < enemyLevelParams.number; i += 1) {
                    const trajectory = new Trajectory(
                        enemyLevelParams.trajectoryPoints,
                        i * NEXT_SHIP_DELAY
                    );
                    ships.push(new EnemyShip(type, trajectory));
                }
            }
        });
        return ships;
    };

    public getLevelTime = (): number => {
        const levelParams = GameLevels[this.currentLevel];
        return levelParams.time;
    };

    public startLevel = () => {
        this.player = this.initPlayer();
        this.enemies = this.initEnemies();
        this.shots = [];
    };

    public setState = (state: GlobalGameState) => {
        this.state = state;
        this.isGameRunning =
            this.state === GlobalGameState.LevelStarted || this.state === GlobalGameState.Resumed;
    };

    public getState = () => this.state;
}

export default new GameState();
