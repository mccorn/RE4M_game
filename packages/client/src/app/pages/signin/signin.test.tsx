import React, { Route, BrowserRouter, Routes } from 'react-router-dom';
import { create } from 'react-test-renderer';

import { Provider } from 'react-redux';

import Button from '@/app/components/common/button/button';
import { store } from '@/app/store/store';
import Signin from './signin';

describe('signIn page tests', () => {
    test('Button snapshot test', () => {
        const tree = create(<Button click={() => ''}>Hello, Jest!</Button>).toJSON();

        expect(tree).toMatchSnapshot();
    });

    test('Signin page snapshot test', () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <Routes>
                        <Route path="/" element={<Signin />} />
                    </Routes>
                </BrowserRouter>
            </Provider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});

export {};
