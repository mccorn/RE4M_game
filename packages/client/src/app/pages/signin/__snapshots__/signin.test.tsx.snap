// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signIn page tests Button snapshot test 1`] = `
<button
  className=""
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Hello, Jest!
</button>
`;

exports[`signIn page tests Signin page snapshot test 1`] = `
<div
  className="formWrapper"
>
  <form
    className="column withGap"
  >
    <label
      className="inputWrapper column"
    >
      <span
        className="line"
      >
        login
      </span>
      <input
        disabled={false}
        name="login"
        onChange={[Function]}
        placeholder="login"
        type="text"
        value=""
      />
    </label>
    <label
      className="inputWrapper column"
    >
      <span
        className="line"
      >
        password
      </span>
      <input
        disabled={false}
        name="password"
        onChange={[Function]}
        placeholder="password"
        type="text"
        value=""
      />
    </label>
    <button
      className=""
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Signin
    </button>
    <button
      className=""
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Signup
    </button>
  </form>
</div>
`;
